### PR TITLE
drivers/n25q128: driver implementation for the n25q128 serial flash memory

### PR DESCRIPTION
**Pull request by ReneHerthel**
### Contribution description
Initial driver implementation for the [N25Q128](https://datasheet4u.com/datasheet-pdf/NumonyxB.V/N25Q128/pdf.php?id=683085) serial flash memory device, which is also installed on the [IoT-LAB M3 open node](https://doc.riot-os.org/group__boards__iotlab-m3.html).

The driver will support the basic ['Extended SPI'](https://html.alldatasheet.com/html-pdf/353314/NUMONYX/N25Q128/6294/21/N25Q128.html)-protocol and all [instructions](https://html.alldatasheet.com/html-pdf/353314/NUMONYX/N25Q128/23394/78/N25Q128.html), except those intended for the 'Dual I/O SPI'- and 'Quad I/O SPI'-protocol.

### Testing procedure
Check the instructions for functionality. For example, check if the bytes read from the ID-instruction match with the bytes in the data sheet.

### CODE REVIEW

**Comment by ReneHerthel** [comment]
@PeterKietzmann I marked it as 'draft', because I need to implement the last instructions for the memory device.

**Comment by PeterKietzmann** [comment]
@ReneHerthel did you continue with this?

File: drivers/include/n25q128.h [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+ *
+ * @param[in] *dev      A pointer to the configured device.
+ * @param[in] address   The address to read bytes from.
+ * @param[out] *buf     Stores the bytes from the address of the device.
+ * @param[in] buf_len   The length of the *buf.
+ */
+void n25q128_read_data_bytes(n25q128_dev_t *dev, uint32_t addr, uint8_t *buf, size_t len);
+
+/*
+ * @brief   Programs bytes into the memory.
```
Comment (PeterKietzmann) [inline drivers/include/n25q128.h:143]: The description says "bytes" but the function says "page". Why not use similar semantics as done for reading, like: `n25q128_write_bytes`

File: drivers/include/n25q128.h [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+/*
+ * @brief   Read a specific amount of data bytes from an address.
+ *
+ * @param[in] *dev      A pointer to the configured device.
+ * @param[in] address   The address to read bytes from.
+ * @param[out] *buf     Stores the bytes from the address of the device.
+ * @param[in] buf_len   The length of the *buf.
+ */
+void n25q128_read_data_bytes(n25q128_dev_t *dev, uint32_t addr, uint8_t *buf, size_t len);
```
Comment (PeterKietzmann) [inline drivers/include/n25q128.h:140]: Are there other bytes than "data" bytes? If not, I would simply name this function `n25q128_read_bytes`.

File: drivers/include/n25q128.h [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+/*
+ * @brief   Programs bytes into the memory.
+ *
+ * The Page Program (PP) instruction allows bytes to be programmed in the memory
+ * (changing bits from 1 to 0). Before it can be accepted, a Write Enable (WREN)
+ * instruction must previously have been executed. After the Write Enable (WREN)
+ * instruction has been decoded, the device sets the Write Enable Latch (WEL).
```
Comment (PeterKietzmann) [inline drivers/include/n25q128.h:148]: Would it be reasonable to add a convenience function that handles everything needed in order to write data or is it intentionally not in place?

File: drivers/include/n25q128.h [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+ * @brief   Erases all bits (set to 1 (FFh)) inside the chosen sector.
+ *
+ * NOTE: This needs to be done, before programming a sector with new values.
+ *       Because the memory can only set bits back to 1 during erasing. So, once
+ *       a bit is programmed from 1 to 0, it cannot be 'overwritten'.
+ *
+ * @param[in] *dev      A pointer to the configured device.
+ * @param[in] addr      The address of the sector to erase.
+ */
+void n25q128_sector_erase(n25q128_dev_t *dev, uint32_t addr);
```
Comment (PeterKietzmann) [inline drivers/include/n25q128.h:167]: So far you were talking about bytes, pages and sectors (which might be all right) but could you introduce the differences somewhere in the description of this driver?

File: drivers/include/n25q128.h [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+void n25q128_program_erase_resume(n25q128_dev_t *dev);
+
+/*
+ * @brief   Check if there is a write in progress.
+ *
+ * @return  The Write In Progress (WIP) bit set to 1 indicates that the memory
+ *          is busy with a Write Status Register, Program or Erase cycle.
+ *          0 indicates no cycles is in progress.
+ */
+bool n25q128_write_in_progress(n25q128_dev_t *dev);
```
Comment (PeterKietzmann) [inline drivers/include/n25q128.h:199]: Mabybe `n25q128_is_busy`?

File: drivers/include/n25q128.h [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+ * executed. After the Write Enable (WREN) instruction has been decoded, the
+ * device sets the Write Enable Latch (WEL).
+ */
+void n25q128_bulk_erase(n25q128_dev_t *dev);
+
+/*
+ * @brief   Pause the program/erase program on the memory.
+ *
+ */
+void n25q128_program_erase_suspend(n25q128_dev_t *dev);
```
Comment (PeterKietzmann) [inline drivers/include/n25q128.h:183]: Does the chip remember the position? Will it return busy if someone suspended a mass erase and tries to write data immediately after?

File: drivers/n25q128/Makefile [hunk 7]
```diff
@@ -0,0 +1,3 @@
+MODULE = n25q128
```
Comment (PeterKietzmann) [inline drivers/n25q128/Makefile:1]: Not necessary when module name equals folder name.

File: drivers/n25q128/include/n25q128_internal.h [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+#define N25Q128_OPCODE_READ         (0x03)
+
+/*
+ * @brief   Read Data Bytes at Higher Speed
+ */
+#define N25Q128_OPCODE_FAST_READ    (0x0B)
+
+/*
+ * @brief   Dual Output Fast Read
+ * NOTE: Only available in Dual mode, which is not supported yet.
```
Comment (PeterKietzmann) [inline drivers/n25q128/include/n25q128_internal.h:50]: `@note`

File: drivers/n25q128/include/n25q128_internal.h [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+/*
+ * @brief   Write Volatile Enhanced Configuration Register
+ */
+#define N25Q128_OPCODE_WRVECR       (0x61)
+
+/** }@ Extended SPI OPCODE */
+
+/*
+ * @brief   Legacy SPI Status Register of the N25Q128 device.
+ */
```
Comment (PeterKietzmann) [inline drivers/n25q128/include/n25q128_internal.h:210]: I guess this needs doxygen grouping. Try to build the documentation (`make doc` from RIOT root directory) and you will see all doxygen errors and warnings.

File: drivers/n25q128/n25q128.c [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#include "n25q128.h"
+#include "include/n25q128_internal.h"
+
+/*
+ * @brief   Clamps a specifc value between minimum and maximum.
+ */
+#define CLAMP(x, min, max)    ((x > max) ? max : ((x < min) ? min : x))
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:34]: Is "clamping" a known term for this?

File: drivers/n25q128/n25q128.c [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+
+/*
+ * @brief   Calls the spi driver for receiving bytes.
+ */
+static inline void _receive(spi_t bus, uint8_t * in, size_t len)
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:39]: Why do you need to wrap the SPI call and why using `SPI_CS_UNDEF`? You don't have to handle CS manually.

File: drivers/n25q128/n25q128.c [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+{
+    spi_transfer_bytes(bus, SPI_CS_UNDEF, false, NULL, in, len);
+}
+
+/*
+ * @brief   Calls the spi driver for sending bytes.
+ */
+static inline void _send(spi_t bus, uint8_t * out, size_t len)
+{
+    spi_transfer_bytes(bus, SPI_CS_UNDEF, false, out, NULL, len);
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:49]: Same as above...

File: drivers/n25q128/n25q128.c [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+/*
+ * @brief   Execute the assembled command of the configured device with the
+ *          prepared instruction set.
+ */
+static inline void _execute(n25q128_dev_t *dev)
+{
+    /* Select the configured memory device, via the Chip-Select-GPIO. */
+    gpio_clear(dev->conf.cs);
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:82]: Not needed (see above)

File: drivers/n25q128/n25q128.c [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+/*
+ * @brief   Read and return the content of the Status Register.
+ */
+static inline uint8_t _read_status_reg(n25q128_dev_t *dev)
+{
+    static uint8_t buf[1] = {0};
+
+    dev->cmd.code = N25Q128_OPCODE_RDSR;
+    dev->cmd.addr = 0;
+    dev->data.buf = buf;
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:145]: `buf` equals zero, why allocating it above?

File: drivers/n25q128/n25q128.c [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+    dev->data.len = 1;
+    dev->opt.mask = N25Q128_OPT_RECV_EN;
+
+    _execute(dev);
+
+    return buf[0];
+}
+
+static inline bool _write_enable_latch(n25q128_dev_t *dev)
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:154]: Could you add a note what the expected return values mean?

File: drivers/n25q128/n25q128.c [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+
+    return N25Q128_OK;
+}
+
+void n25q128_read_id(n25q128_dev_t *dev, uint8_t *buf, size_t len)
+{
+    dev->cmd.code = N25Q128_OPCODE_RDID;
+    dev->cmd.addr = 0;
+    dev->data.buf = buf;
+    dev->data.len = CLAMP(len, 1, 20);
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:212]: This would not be needed if the type was defined.

File: drivers/n25q128/n25q128.c [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+
+static inline bool _write_enable_latch(n25q128_dev_t *dev)
+{
+    uint8_t spi_status_reg = _read_status_reg(dev);
+    return (spi_status_reg & N25Q128_STAT_REG_WEL);
+}
+
+int n25q128_init(n25q128_dev_t *dev)
+{
+    int status = N25Q128_OK;
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:162]: Add asserts in case `dev` is not set

File: drivers/n25q128/n25q128.c [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+    dev->cmd.addr = 0;
+    dev->data.buf = buf;
+    dev->data.len = CLAMP(len, 1, 20);
+    dev->opt.mask = N25Q128_OPT_RECV_EN;
+
+    _execute(dev);
+}
+
+void n25q128_read_data_bytes(n25q128_dev_t *dev, uint32_t addr, uint8_t *buf, size_t len)
+{
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:219]: assert on `buf` not set as well.

File: drivers/n25q128/n25q128.c [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+    dev->cmd.code = N25Q128_OPCODE_READ;
+    dev->cmd.addr = addr;
+    dev->data.buf = buf;
+    dev->data.len = CLAMP(len, 1, len); /* 'infinite' bytes possible */
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:223]: With assertion above, you could simply pass `len` here I guess.

File: drivers/n25q128/n25q128.c [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+    uint8_t set[4] = {0};
+
+    /* Possible length is 4 Byte (with address) or 1 Byte (only instruction). */
+    size_t len = (dev->opt.enabled.addr) ? 4 : 1;
+
+    /* The first byte will allways be the operation code. */
+    set[0] = (dev->cmd.code & 0xFF);
+
+    /* The address is made of 3 Bytes, which needs to be shifted correctly. */
+    set[1] = ((dev->cmd.addr & 0x00FF0000) >> 16); // = 0x000000FF
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:67]: Don't use `//` comment style

File: drivers/n25q128/include/n25q128_internal.h [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+
+/*
+ * @brief   Dual Output Fast Read
+ * NOTE: Only available in Dual mode, which is not supported yet.
+ */
+#define N25Q128_OPCODE_DOFR         (0x3B)
+
+/*
+ * @brief   Dual Input/Output Fast Read
+ * NOTE: Only available in Dual mode, which is not supported yet.
```
Comment (PeterKietzmann) [inline drivers/n25q128/include/n25q128_internal.h:56]: `@note`

File: drivers/n25q128/include/n25q128_internal.h [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+
+/*
+ * @brief   Write Volatile Enhanced Configuration Register
+ */
+#define N25Q128_OPCODE_WRVECR       (0x61)
+
+/** }@ Extended SPI OPCODE */
+
+/*
+ * @brief   Legacy SPI Status Register of the N25Q128 device.
```
Comment (PeterKietzmann) [inline drivers/n25q128/include/n25q128_internal.h:209]: Group needed. Build doxygen by `make doc` from RIOT root directory to see all doxygen errors and warnings.

File: drivers/n25q128/n25q128.c (hunk 10 above)
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:34]: Is "clamp" a known term?

File: drivers/n25q128/n25q128.c [hunk 23]
```diff
@@ -0,0 +1,293 @@
...
+ * @brief   Clamps a specifc value between minimum and maximum.
+ */
+#define CLAMP(x, min, max)    ((x > max) ? max : ((x < min) ? min : x))
+
+/*
+ * @brief   Calls the spi driver for receiving bytes.
+ */
+static inline void _receive(spi_t bus, uint8_t * in, size_t len)
+{
+    spi_transfer_bytes(bus, SPI_CS_UNDEF, false, NULL, in, len);
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:41]: Why do you need to wrap around the SPI call? And why `SPI_CS_UNDEF`? You don't need to handle CS yourself.

File: drivers/n25q128/n25q128.c (hunk 12 above)
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:49]: Same as above.

File: drivers/n25q128/n25q128.c (hunk 13 above)
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:82]: No need to handle manually.

File: drivers/n25q128/n25q128.c (hunk 15 above)
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:154]: Would you spend a comment what to expect in return?

File: drivers/n25q128/n25q128.c (hunk 17 above)
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:162]: Assert if `dev` not set

File: drivers/n25q128/n25q128.c (hunk 16 above)
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:212]: Not needed if proper type was in place.

File: drivers/n25q128/n25q128.c [hunk 24]
```diff
@@ -0,0 +1,293 @@
...
+    dev->data.buf = buf;
+    dev->data.len = CLAMP(len, 1, 20);
+    dev->opt.mask = N25Q128_OPT_RECV_EN;
+
+    _execute(dev);
+}
+
+void n25q128_read_data_bytes(n25q128_dev_t *dev, uint32_t addr, uint8_t *buf, size_t len)
+{
+    dev->cmd.code = N25Q128_OPCODE_READ;
```
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:220]: Assert if dev or buf not set.

File: drivers/n25q128/n25q128.c (hunk 19 above)
Comment (PeterKietzmann) [inline drivers/n25q128/n25q128.c:223]: With above assert, you could pass `len` directly.

File: tests/driver_n25q128/README.md [hunk 25]
```diff
@@ -0,0 +1 @@
+TODO
```
Comment (PeterKietzmann) [inline tests/driver_n25q128/README.md:1]: Please do :)

File: tests/driver_n25q128/main.c [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+static int cmd_init(int argc, char **argv)
+{
+    (void) argc;
+    (void) argv;
+    int status = 0;
+
+    puts("cmd_init: Initialize the N25Q128 flash memory.\n");
+
+    // XXX: Specific configuration for iotlab-m3 nodes.
+    n25q128.conf.bus = EXTFLASH_SPI;
```
Comment (PeterKietzmann) [inline tests/driver_n25q128/main.c:78]: Configuration should be done as described in the [drivers guide](https://doc.riot-os.org/driver-guide.html#driver-guide-config)

File: tests/driver_n25q128/main.c [hunk 27]
```diff
@@ -0,0 +1,254 @@
...
+    n25q128.conf.mode = SPI_MODE_0;
+    n25q128.conf.clk = SPI_CLK_100KHZ;
+    n25q128.conf.cs = EXTFLASH_CS;
+    n25q128.conf.write = EXTFLASH_WRITE;
+    n25q128.conf.hold = EXTFLASH_HOLD;
+
+    status = n25q128_init(&n25q128);
```
Comment (PeterKietzmann) [inline tests/driver_n25q128/main.c:85]: Should be implemented in `auto_init`

**Comment by PeterKietzmann** [comment]
The driver should be implemented against a generic interface. Did you consider [Serial NOR flash](http://doc.riot-os.org/group__drivers__mtd__spi__nor.html)?

**Pull request closed without merging**